dartsim/src/SimulationFeatures.cc

```diff
@@ -52,7 +64,57 @@ void SimulationFeatures::WorldForwardStep(
 
   // TODO(MXG): Parse input
   world->step();
-  // TODO(MXG): Fill in output and state
+  this->WriteRequiredData(_h);
+  this->Write(_h.Get<JointPositions>());
+  // TODO(MXG): Fill in state
+}
+
+void SimulationFeatures::Write(JointPositions &/*_positions*/) const
+{
+  // TODO(adlarkin) implement this, if it's needed?
+}
+
+void SimulationFeatures::Write(WorldPoses &_poses) const
+{
+  // remove link poses from the previous iteration
+  _poses.entries.clear();
+
+  std::unordered_map<std::size_t, math::Pose3d> newPoses;
+
+  for (const auto &link : this->links.idToObject)
+  {
+    const auto id = link.first;
+    const auto info = link.second;
+
+    // make sure the link exists
+    if (info && info->link)
+    {
+      WorldPose wp;
+      wp.pose = ignition::math::eigen3::convert(
+          info->link->getWorldTransform());
+      wp.pose.Pos() = ignition::math::eigen3::convert(
+          info->link->getCOM());
+      wp.body = id;
+
+      // if the link's pose is new or has changed,
+      // add the link to the output poses
+      auto iter = this->prevLinkPoses.find(id);
+      if ((iter == this->prevLinkPoses.end()) ||
+          !iter->second.Pos().Equal(wp.pose.Pos(), 1e-6) ||
+          !iter->second.Rot().Equal(wp.pose.Rot(), 1e-6))
+      {
+        _poses.entries.push_back(wp);
+      }
+
+      newPoses[id] = wp.pose;
```

Comment:
The reason why I left this outside of the `if` block is because of floating point precision. My thought was that I should save the most recent pose in `newPoses` no matter what each time, since there's a small tolerance (`1e-6`) in the pose equality check. Since `newPoses` is used as `prevLinkPoses` the next time around, this ensures that we are comparing the newest link pose to the actual previous pose, instead of to a pose that is _very close_ to the previous pose. Does that make sense? If you think that leaving this outside of the `if` block is okay, I can add a comment there that mentions what I just described to you. Otherwise, I can move this to be inside of the `if` block.

Reply:
I would think you'd want to compare it to the last pose sent to the user though because you want to err on the side of sending too much to the user (eg. ign-gazebo). Consider, for example, if each successive time the pose changes by +1e-7, since you always compare against the previous actual pose, it will always look like there no new pose to send to the user. Over time, this could cause a significant amount of drift.

Reply:
You're definitely correct. I believe that this should now be fixed in https://github.com/ignitionrobotics/ign-physics/pull/223/commits/494f6c567926e42805a94d29adb65e66cc0c8186